File: mcp-code-intelligence-nodejs/src/memory/role-filter.ts
/**
 * RoleFilter — maps agent roles to relevant knowledge entry types.
 */

const ROLE_TYPES: Record<string, Set<string>> = {
  DEV: new Set(['CODE_ENTITY', 'ARCHITECTURE', 'API_DESIGN', 'DECISION']),
  BA: new Set(['REQUIREMENT', 'CONTEXT', 'DECISION', 'LESSON_LEARNED']),
  QA: new Set(['PROCEDURE', 'REQUIREMENT', 'ERROR_PATTERN', 'LESSON_LEARNED']),
  SA: new Set(['ARCHITECTURE', 'API_DESIGN', 'CODE_ENTITY', 'DECISION']),
  DEVOPS: new Set(['PROCEDURE', 'ARCHITECTURE', 'CONTEXT']),
};

/** Get allowed types for a role. Returns null if no filter (all types). */
export function typesForRole(role?: string | null): Set<string> | null {
  if (!role || !role.trim()) return null;
  return ROLE_TYPES[role.toUpperCase()] ?? null;
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/role_filter.py
"""RoleFilter — maps agent roles to relevant knowledge entry types."""

from __future__ import annotations

ROLE_TYPES: dict[str, set[str]] = {
    "DEV": {"CODE_ENTITY", "ARCHITECTURE", "API_DESIGN", "DECISION"},
    "BA": {"REQUIREMENT", "CONTEXT", "DECISION", "LESSON_LEARNED"},
    "QA": {"PROCEDURE", "REQUIREMENT", "ERROR_PATTERN", "LESSON_LEARNED"},
    "SA": {"ARCHITECTURE", "API_DESIGN", "CODE_ENTITY", "DECISION"},
    "DEVOPS": {"PROCEDURE", "ARCHITECTURE", "CONTEXT"},
}


def types_for_role(role: str | None) -> set[str] | None:
    """Get allowed types for a role. Returns None if no filter (all types)."""
    if not role or not role.strip():
        return None
    return ROLE_TYPES.get(role.upper())
